src/protocols/secure_channel/CASESession.cpp

```diff
@@ -639,47 +710,81 @@ CHIP_ERROR CASESession::HandleSigmaR2(const System::PacketBufferHandle & msg)
-                              kP256_PublicKey_Length - kTAGSize,
-                          nullptr, 0, tag, kTAGSize, sr2k, kAEADKeySize, kIVSR2, kIVLength, msg_r2_encrypted);
+    err = AES_CCM_decrypt(msg_R2_Encrypted->Start(), msg_R2_Encrypted->DataLength(), nullptr, 0, tag, kTAGSize, sr2k, kAEADKeySize,
+                          kIVSR2, kIVLength, msg_R2_Encrypted->Start());
+    SuccessOrExit(err);
+
+    tlvReader.Init(std::move(msg_R2_Encrypted));
+    containerType = TLV::kTLVType_Structure;
+    err           = tlvReader.Next(containerType, TLV::AnonymousTag);
+    SuccessOrExit(err);
+    err = tlvReader.EnterContainer(containerType);
+    SuccessOrExit(err);
+
+    err = tlvReader.FindElementWithTag(CASETLVTag::kNOC, suppTlvReader);
+    SuccessOrExit(err);
+    VerifyOrExit(suppTlvReader.GetType() == TLV::kTLVType_UTF8String || suppTlvReader.GetType() == TLV::kTLVType_ByteString,
+                 err = CHIP_ERROR_WRONG_TLV_TYPE);
+    responderOpCertLen = static_cast<uint16_t>(suppTlvReader.GetLength());
+    err                = suppTlvReader.GetBytes(responderOpCert, responderOpCertLen);
     SuccessOrExit(err);
 
     // Step 5
     // Validate responder identity located in msg_r2_encrypted
     // Constructing responder identity
-    err = Validate_and_RetrieveResponderID(&buf, remoteCredential, &remoteDeviceOpCert, remoteDeviceOpCertLen);
+    err = Validate_and_RetrieveResponderID(responderOpCert, responderOpCertLen, remoteCredential);
     SuccessOrExit(err);
 
     // Step 6 - Construct msg_R2_Signed and validate the signature in msg_r2_encrypted
-    msg_r2_signed_len = static_cast<uint16_t>(sizeof(uint16_t) + remoteDeviceOpCertLen + kP256_PublicKey_Length * 2);
+    msg_r2_signed_len = static_cast<uint16_t>(sizeof(uint16_t) + responderOpCertLen + kP256_PublicKey_Length * 2);
 
-    VerifyOrExit(msg_R2_Signed.Alloc(msg_r2_signed_len), err = CHIP_ERROR_NO_MEMORY);
+    msg_R2_Signed = System::PacketBufferHandle::New(msg_r2_signed_len + System::PacketBuffer::kDefaultHeaderReserve);
+    VerifyOrExit(!msg_R2_Signed.IsNull(), err = CHIP_SYSTEM_ERROR_NO_MEMORY);
 
-    sigLen = buflen - kSigmaParamRandomNumberSize - sizeof(encryptionKeyId) - kTrustedRootIdSize - kP256_PublicKey_Length -
-        sizeof(uint16_t) - remoteDeviceOpCertLen - kTAGSize;
-    {
-        MutableByteSpan msg_R2_Span(msg_R2_Signed.Get(), msg_r2_signed_len);
-        err = ConstructSignedCredentials(&buf, remoteDeviceOpCert, remoteDeviceOpCertLen, msg_R2_Span, sigmaR2SignedData, sigLen);
-        SuccessOrExit(err);
-    }
+    err = ConstructTBS2Data(responderOpCert, responderOpCertLen, msg_R2_Signed);
+    SuccessOrExit(err);
+
+    err = tlvReader.FindElementWithTag(CASETLVTag::kSignature, suppTlvReader);
+    SuccessOrExit(err);
+    VerifyOrExit(suppTlvReader.GetType() == TLV::kTLVType_UTF8String || suppTlvReader.GetType() == TLV::kTLVType_ByteString,
```

Comment:
```suggestion
    VerifyOrExit(suppTlvReader.GetType() == TLV::kTLVType_ByteString,
```

According to the spec the signature SHALL be encoded as OCTET STRING. Why UTF8String is also defined as an option here? 

Similar comment applies to many other placed in this PR.

Reply:
Updated.